### App Profiles For the App Store and Google Play Markets 
- Introduction
    - Our aim is to help our developers understand what type apps are likely to attract more users on google play and the Apple store.
    - we are only interested in English apps and the apps which are free of cost.

### Information about the datset
The dataset is about various mobile apps in ios App store.
The dataset is taken from kaggle.
The [documentation](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps)

| Column name | Description             | 
|---          |---                      |
|"id"         | App ID                  |
|"track_name" | App Name                |
|"size_bytes" | Memory size (in Bytes)  |
|"app_desc"   | Application description |

In [1]:
import pandas as pd
from csv import reader

Below is a function named  'explore_data' which helps us to get the information about the datasets. 
For example, printing some rows, displaying number of rows and column.
The function assumes that there is no header in the dataset, so while using the function the starting row numberis '1'.



In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [3]:
applestore=open("AppleStore.csv")
applestore = reader(applestore)
applestore_full= list(applestore)
ios_header=applestore_full[0]
applestore=applestore_full[1:]


google_play_store=open("googleplaystore.csv")
google_play_store=reader(google_play_store)
android_full=list(google_play_store)
android_header=android_full[0]
android=android_full[1:]


In [4]:
explore_data(applestore,1,4,True)

['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


Number of rows: 7197
Number of columns: 16


In [5]:
explore_data(android,1,4,True)

['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 10841
Number of columns: 13


In [6]:
# Extracting the column names of both the datasets
print(ios_header)
print('\n')
print(android_header)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


## Deleting wrong data 

The google play dataset has a [discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion) section. And [one of the discussion says](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015), the 10472(header not included) row has an error.


In [7]:
# The android dataset has a wrong row. It is 10472th row if header is not included
print(android_header)
print('\n')
print(android[10472])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


Indeed the row has an error, it has 'category' column missing, as a result the values of the corresponding columns have been shifted right and it has 'genres' column as an empty string.

In [8]:
# Deleting the wrong row 10472
# Be aware of not running this code more than once.
print(len(android))
del android[10472]
print(len(android))

10841
10840


## Removing the duplicate rows

If we explore the Google Play data set long enough, we'll find that some apps have more than one entry.

- We write a code to check the unique apps in android apps dataset.
- Then we print few duplicate apps.

In [9]:
unique_apps=[]
duplicate_apps=[]
for apps in android:
    if apps[0] not in unique_apps:
        unique_apps.append(apps[0])
    else:
        duplicate_apps.append(apps[0])
        
print(duplicate_apps[862:868])
print('\n')
print('Duplicate apps: ',len(duplicate_apps))
print('Unique apps: ',len(unique_apps))

['Gboard - the Google Keyboard', 'Granny', 'Google', 'Hangouts', 'G Cloud Backup', 'Google Drive']


Duplicate apps:  1181
Unique apps:  9659


In [10]:
for apps in android:
    name=apps[0]
    if name=='Google':
        print(apps)

['Google', 'TOOLS', '4.4', '8033493', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 3, 2018', 'Varies with device', 'Varies with device']
['Google', 'TOOLS', '4.4', '8021623', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 3, 2018', 'Varies with device', 'Varies with device']


## Criterion for removing duplicate entries
- By examining the apps that have duplicate entries in our android apps dataset we find that they all have similar values with repect to the column values. The column values where they differs sometimes is in their number of user ratings('Rating').
- So, we will keep the unique entry of the apps which has highest number of user ratings beacuse it gives the most recent updated data(In case of duplicate entries of apps). 

Creating a dictionary where each key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app.

The dictionary will contain the unique apps names and as we got from previuos code cell that the number of unique apps in google play store dataset is 9659.

In [11]:
reviews_max={}
for apps in android:
    name=apps[0]
    n_reviews=float(apps[3])
    if name in reviews_max:
        if reviews_max[name] < n_reviews:
            reviews_max[name]=n_reviews
    else:
        reviews_max[name]=n_reviews
# Checking the length of the dictionary is 9659 or not.   
print('Expected length: ',len(android)-1181)
print('Actual length: ',len(reviews_max))

Expected length:  9659
Actual length:  9659


- Now, we will create a clean android apps dataset which wil ontain only the unique apps and their information.
- We will use the previously created unique apps with highest number of reviews dictionary for this purpose.
- we will create two lists, 'android_clean' which is ourclean android apps dataset and 'already_added' to check is the app is already added to our cleaned dataset or not.

In [12]:
android_clean=[]
already_added=[]

for apps in android:
    name=apps[0]
    n_reviews=float(apps[3])
    if (n_reviews== reviews_max[name]) and (name not in already_added):
        android_clean.append(apps)
        already_added.append(name)


Now let's quickly explore the new data set, and confirm that the number of rows is 9,659.

In [13]:
explore_data(android_clean,0,4,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


Number of rows: 9659
Number of columns: 13


So, everything went as expected. Untill now everything is fine. 

## Removing Non English Apps
- we use English for the apps we develop at our company, and we'd like to analyze only the apps that are directed toward an English-speaking audience.
- We're not interested in keeping these apps, so we'll remove them. One way to go about this is to remove each app with a name containing a symbol that is not commonly used in English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;), and other symbols (+, *, /).
- All these characters that are specific to English texts are encoded using the ASCII standard. Each ASCII character has a corresponding number between 0 and 127 associated with it, and we can take advantage of that to build a function that checks an app name and tells us whether it contains non-ASCII characters.
- We can use the built-in function ord() to check the encoding of a character.

### Part 1:
Analyzing the encoding of characters and checking if it's an English string or not. 
- we first build a function that takes in a string and returns False if there's any character in the string that doesn't belong to the set of common English characters, otherwise it returns True.
- The function checks if any character present in the string has encoding value greater than 127 or not. If it has an encoding value greater than 127 then it contains character outside the common ASCII range of 127 an may be a non-English app. 

In [14]:
def is_English(string):
    for character in string:
        if ord(character) > 127:
            return False
    return True

print(is_English('Instagram'))
print(is_English('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_English('Docs To Go™ Free Office Suite'))
print(is_English('Instachat 😜'))

True
False
False
False


Examining the above results we find that the function has some issues, and if we use it to remove non English app then we might remove some potential English apps.
- Some English Application name contain emojis and specila trademark or copyright signs.
- So, if we only check the app name contains normal English texts or not then it might be a big problem.

In [15]:
print(ord('™'))
print(ord('😜'))

8482
128540


### Part2:

If we use our previously cretaed function then we might lose some potential English apps as the function will lebel some English apps as non-English.
- So, we now update our function a bit, so that if an app name contains more than three characters with corresponding numbers falling outside the ASCII range it will lebel the app as non-English.
- Our function is still not perfect, as for example an English app name might contain more than three emojis or trademark signs, copyright signs and emojis all together. But the function is quite effective now.

In [16]:
def is_English(string):
    non_ascii=0
    for character in string:
        if ord(character) > 127:
            non_ascii+=1
    
    if non_ascii > 3:
        return False
    else:
        return True        

In [17]:
print(is_English('Docs To Go™ Free Office Suite'))
print(is_English('Instachat 😜'))
print(is_English('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


Our function now works as we expected.

But still our function is not perfect and we might loose some potential English apps. But it can serve our purpose quite effectively now. Later, we can dive deep to improve this function.

Now, we will use the function 'is_English' to remove non-English apps from both **applestore** and **android_clean** dataset.

In [18]:
applestore_English=[]
android_English=[]

for apps in android_clean:
    name=apps[0]
    if is_English(name):
        android_English.append(apps)
        
for apps in applestore:
    name=apps[1]         # 'track_name' is the name of the ios app and 
    if is_English(name): # it is at column number 1.
        applestore_English.append(apps)
        
print("Exploring new android_English list: ")
print('\n')
explore_data(android_English,0,3, True)
print('\n')
print("Exploring new applestore_English list: ")
print('\n')
explore_data(applestore_English,0,3,True)

Exploring new android_English list: 


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


Exploring new applestore_English list: 


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Cla

So, there are 9614 android apps and 6183 ios apps which falls into English category.

## Taking only the free apps

Recall that we are only concerned with apps that are free to download and install.
- The company for which we are doing the data analysis only build apps that are free to download and install.
- Thus we have to consider only the apps that have price as Free or '0'.
- Our dataset contains both the free apps and non-free apps.
- so, we isolate only the free apps and store them in **applestore_final** and **android_final** lists.
- In case of free ios apps the price is ** '0.0' ** and in case of free android apps the price is ** '0'**.
- The price column in ios apps dataset is **4** and android apps dataset is **7**.

In [19]:
applestore_final=[]
android_final=[]

for apps in android_English:
    price=apps[7]
    if price == '0':
        android_final.append(apps)
        
for apps in applestore_English:
    price=apps[4]
    if price == '0.0':
        applestore_final.append(apps)
        
print("Exploring final android_final list: ")
print('\n')
explore_data(android_final,0,3, True)
print('\n')
print("Exploring final applestore_final list: ")
print('\n')
explore_data(applestore_final,0,3,True)

Exploring final android_final list: 


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 8864
Number of columns: 13


Exploring final applestore_final list: 


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Cla

Finally we have 8864 Free and English android apps and 3222 Free and English ios  apps.

## Now our actual analysis on the datasets begin

Let us print the headers of both the datsets again

In [20]:
print('ios dataset column names: ', ios_header)
print('\n')
print('google play store dataset column names: ', android_header)

ios dataset column names:  ['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


google play store dataset column names:  ['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


## Most common App by Genre

### Part 1

As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

1) Build a minimal Android version of the app, and add it to Google Play.

2) If the app has a good response from users, we then develop it further.

3) If the app is profitable after six months, we also build an iOS version of the app and add it to the App Store.


Because our end goal is to add the app on both the App Store and Google Play, we need to find app profiles that are successful on both markets. For instance, a profile that might work well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by getting a sense of the most common genres for each market. For this, we'll build a frequency table for the prime_genre column of the App Store data set, and the Genres and Category columns of the Google Play data set.

### Part 2

We'll build two functions we can use to analyze the frequency tables:
- One function to generate frequency tables that show percentages.
- Another function we can use to display the percentages in a descending order.
- App Store dataset is named as **applestore_final** and Play Store dataset is named as ** android_final**. 

In [21]:
def freq_table(dataset,index):
    length=0
    table={}
    for row in dataset:
        length+=1
        genre=row[index]
        if genre in table:
            table[genre]+=1
        else:
            table[genre]=1
            
    table_percentages={}
    for element in table:
        table_percentages[element]= (table[element]/length)*100
    
    return table_percentages

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

### Part 3

Now, we generate frequency table for the prime_genre column of the App Store data set.
- ** prime_genre** column is at index **11**.

In [22]:
display_table(applestore_final,11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


Looking into the frequency table we find that more than 50% of the apps in app store is Games (58.16 %), then Entertainment apps are 7.88 %, followed by  Photo & Video apps nearly 5%, Education apps (3.66 %) and Social Networking apps (3.29 %).
We did the analysis on English and free apps. So, atleast in the case of Free and English apps more than 50 % of the apps are for fun and enjoyment in app store.

Let's continue by examining the Genres and Category columns of the Google Play data set (two columns which seem to be related).
- First, we examine the **category** column in android apps dataset and it has column index **1**.
- Second, we examine the **Genres** column in android apps dataset (**android_final**) which has a column index **-4**.

In [23]:
display_table(android,1) # CATEGORY

FAMILY : 18.19188191881919
GAME : 10.55350553505535
TOOLS : 7.776752767527675
MEDICAL : 4.271217712177122
BUSINESS : 4.243542435424354
PRODUCTIVITY : 3.911439114391144
PERSONALIZATION : 3.616236162361624
COMMUNICATION : 3.5701107011070112
SPORTS : 3.5424354243542435
LIFESTYLE : 3.5239852398523985
FINANCE : 3.3763837638376386
HEALTH_AND_FITNESS : 3.1457564575645756
PHOTOGRAPHY : 3.0904059040590406
SOCIAL : 2.7214022140221403
NEWS_AND_MAGAZINES : 2.61070110701107
SHOPPING : 2.3985239852398523
TRAVEL_AND_LOCAL : 2.3800738007380073
DATING : 2.158671586715867
BOOKS_AND_REFERENCE : 2.1309963099630997
VIDEO_PLAYERS : 1.6143911439114391
EDUCATION : 1.4391143911439115
ENTERTAINMENT : 1.3745387453874538
MAPS_AND_NAVIGATION : 1.2638376383763839
FOOD_AND_DRINK : 1.1715867158671587
HOUSE_AND_HOME : 0.8118081180811807
LIBRARIES_AND_DEMO : 0.7841328413284132
AUTO_AND_VEHICLES : 0.7841328413284132
WEATHER : 0.7564575645756457
ART_AND_DESIGN : 0.5996309963099631
EVENTS : 0.5904059040590406
PARENTING : 

The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.

In [24]:
display_table(android_final,-4) # Genres

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075


The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

## Most Popular Apps by Genre on the App Store

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the _Installs_ column, but this information is missing for the App Store data set. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

Below, we calculate the average number of user ratings per app genre on the App Store:
- The **rating_count_tot** column in ios dataset has index **5**.

In [25]:
appstore_genres=freq_table(applestore_final,11)
app_user_ratings=[]
for genre in appstore_genres:
    total=0           #stores the sum of user ratings specific to each genre.
    len_genre=0       #Stores number of apps specific to each genre.
    for apps in applestore_final:
        genre_app=apps[11]
        if genre_app == genre:
            number= float(apps[5])
            total+=number
            len_genre+=1
    average_no_ratings=total/len_genre
    genre_rating_tuple=(average_no_ratings,genre)
    app_user_ratings.append(genre_rating_tuple)

app_user_ratings=sorted(app_user_ratings,reverse=True)
for apps in app_user_ratings:
    print(apps[1]," : ",apps[0])
    
    

Navigation  :  86090.33333333333
Reference  :  74942.11111111111
Social Networking  :  71548.34905660378
Music  :  57326.530303030304
Weather  :  52279.892857142855
Book  :  39758.5
Food & Drink  :  33333.92307692308
Finance  :  31467.944444444445
Photo & Video  :  28441.54375
Travel  :  28243.8
Shopping  :  26919.690476190477
Health & Fitness  :  23298.015384615384
Sports  :  23008.898550724636
Games  :  22788.6696905016
News  :  21248.023255813954
Productivity  :  21028.410714285714
Utilities  :  18684.456790123455
Lifestyle  :  16485.764705882353
Entertainment  :  14029.830708661417
Business  :  7491.117647058823
Education  :  7003.983050847458
Catalogs  :  4004.0
Medical  :  612.0


## Most Popular Apps by Genre on the Play Store

For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):


In [26]:
display_table(android_final,5) #Installs

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category).
- The **Category** column has index **1** and the ** Installs** column has index **5**.

In [33]:
playstore_genres=freq_table(android_final,1)
app_installs=[]
for category in playstore_genres:
    total =0  # stores the sum of installs specific to each genre.
    len_category=0 #store the number of apps specific to each genre.
    for row in android_final:
        category_app=row[1]
        if category_app == category:
            number= row[5]
            number=number.replace(',', '')
            number=number.replace('+', '')
            number=float(number)
            total+=number
            len_category+=1
    average_installs= total/len_category
    installs_tuple=(average_installs,category)
    app_installs.append(installs_tuple)

app_installs=sorted(app_installs,reverse=True)   
for apps in app_installs:
    print(apps[1]," : ",apps[0])

COMMUNICATION  :  38456119.167247385
VIDEO_PLAYERS  :  24727872.452830188
SOCIAL  :  23253652.127118643
PHOTOGRAPHY  :  17840110.40229885
PRODUCTIVITY  :  16787331.344927534
GAME  :  15588015.603248259
TRAVEL_AND_LOCAL  :  13984077.710144928
ENTERTAINMENT  :  11640705.88235294
TOOLS  :  10801391.298666667
NEWS_AND_MAGAZINES  :  9549178.467741935
BOOKS_AND_REFERENCE  :  8767811.894736841
SHOPPING  :  7036877.311557789
PERSONALIZATION  :  5201482.6122448975
WEATHER  :  5074486.197183099
HEALTH_AND_FITNESS  :  4188821.9853479853
MAPS_AND_NAVIGATION  :  4056941.7741935486
FAMILY  :  3695641.8198090694
SPORTS  :  3638640.1428571427
ART_AND_DESIGN  :  1986335.0877192982
FOOD_AND_DRINK  :  1924897.7363636363
EDUCATION  :  1833495.145631068
BUSINESS  :  1712290.1474201474
LIFESTYLE  :  1437816.2687861272
FINANCE  :  1387692.475609756
HOUSE_AND_HOME  :  1331540.5616438356
DATING  :  854028.8303030303
COMICS  :  817657.2727272727
AUTO_AND_VEHICLES  :  647317.8170731707
LIBRARIES_AND_DEMO  :  638